integrated_channels/canvas/client.py

```diff
@@ -378,37 +378,65 @@ def _handle_canvas_assignment_retrieval(
             course_id (str) : the Canvas course ID relating to the course which the client is currently
             transmitting learner data to.
         """
-        # First, check if the course assignment already exists
+        # Check if the course assignment already exists
         canvas_assignments_url = '{canvas_base_url}/api/v1/courses/{course_id}/assignments'.format(
             canvas_base_url=self.enterprise_configuration.canvas_base_url,
             course_id=course_id
         )
         resp = self.session.get(canvas_assignments_url)
 
-        if resp.status_code >= 400:
-            raise ClientError(
-                "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
-                    resp.text,
-                ),
-                resp.status_code
-            )
+        more_pages_present = True
+        current_page_count = 0
+        assignment_id = ''
 
-        assignments_resp = resp.json()
-        assignment_id = None
-        for assignment in assignments_resp:
-            try:
-                if assignment['integration_id'] == integration_id:
-                    assignment_id = assignment['id']
-                    break
-            # The validation check above should ensure that we have a 200 response from Canvas, but sanity catch if we
-            # have a unexpected response format
-            except (KeyError, ValueError, TypeError) as error:
+        # current_page_count serves as a timeout, limiting to a max of 150 pages of requests
+        while more_pages_present and current_page_count < 150:
```

Comment:
I now realize this is assignments for a single course right? so it say 2000 students are enrolled in the course over time, we expect to get say 2000 assignments in this response total? if per_page=10, we are going to possibly make 200 calls.... not sure if these numbers are realistic but now I am like if we keep per_page=10 we could end up making a lot of requests, assuming large enrollment counts

Reply:
The assignments are per subsection on edX, we shouldn't be creating dups per student.

Reply:
@binodpant this ends up being fine because the API doesn't work the way you expect (I also had to look this up because I too was expecting the API call to need a student_id filter or similar on it). Instead, an 'assignment' object here is an object that belongs to a course, there is a different API (not the one we call) called 'List assignments for user' that does what we would expect would happen.

Reply:
thanks if we’re not doing crazy requests/loops per course we should be fine, we can also collect data. I also wish canvas had better search apis so we coukd look for specific assignments for presence but ok...let’s sendcthis out so we resolve the duplicate issue in the least